velox/functions/prestosql/aggregates/ApproxPercentileAggregate.cpp

```diff
@@ -67,42 +69,69 @@ struct KllSketchAccumulator {
     }
   }
 
-  void append(const typename KllSketch<T>::View& view) {
+  void append(const KllView<T>& view) {
     sketch_.mergeViews(folly::Range(&view, 1));
   }
 
-  void append(const std::vector<typename KllSketch<T>::View>& views) {
+  void append(const std::vector<KllView<T>>& views) {
     sketch_.mergeViews(views);
   }
 
-  void finalize() {
+  // Creates a copy of the KllSketch, merges the largeCountValues_ into it,
+  // compacts it, and returns it.
+  // Makes a copy so that uses std::allocator so that this is safe to call
+  // during spilling which may run in parallel.  HashStringAllocator is not
+  // thread safe, so merging into/compacting the original KllSketch which
+  // depends on it can lead to concurrency bugs.
+  functions::kll::KllSketch<T> compact() {
+    functions::kll::KllSketch<T> newSketch =
+        functions::kll::KllSketch<T>::fromView(
+            sketch_.toView(), std::allocator<T>(), random::getSeed());
+
     if (!largeCountValues_.empty()) {
-      flush();
+      std::vector<functions::kll::KllSketch<T>> sketches;
```

Comment:
Can extract this duplicated code to a private function taking in allocator as parameter?  Also we should clear `largeCountValues_` here as well.

Reply:
> Can extract this duplicated code to a private function taking in allocator as parameter?
Good idea! i'll do that.

> Also we should clear largeCountValues_ here as well.
The reason I didn't clear largeCountValues_ is because we're merging them into a temporary KllSketch.  Clearing the largeCountValues_ would be a destructive operation that leaves the KllSketchAccumulator in an incomplete state.  Although I don't think anyone would use the KllSketchAccumulator after compaction is called currently, it could quietly lead to incorrect results in the future if anyone does.  The risk is low but the damage seems high, and given the Accumulator will probably get destructed soon after (because of the spill or flushing the results) the benefit seems low (free a little bit of memory a little sooner) I don't think it's worth it.

Reply:
OK then we leave the `largeCountValues_` in this function

Reply:
Thanks, will do.